README.md
# tainted-lua

tainted-lua is a customized Lua 5.1 implementation extending the virtual machine with a tainted execution model based off the World of Warcraft user interface environment.

A high-level overview of how the tainted execution model works can be boiled down to these three points.

- Any untrusted (or insecure) code that writes values will "taint" whatever is being stored.
- Any trusted (or secure) code that reads tainted values will spread that taint to the current call stack.
- Any function can check whether or not the call stack is tainted to determine whether or not a privileged action should be permitted.

This project is intended to serve as a close approximation of the tainted execution model against a reference client environment, and should be accurate enough for any legitimate addon development and external tooling use cases. This project is **not** a binary-compatible implementation of taint as found within the actual game client itself, and does not aim to be fully bug-compatible with the reference client with respect to any potential security issues that can impact the live game service. Any security issues found within the reference client during development of this project have - and continue to be - submitted to Blizzard for resolution.

## Building

CMake is used for building the project. The presets defined in `CMakePresets.json` at the root of the repository represent the supported build configurations that are tested by the CI; these are Linux (GCC), macOS (Clang) and Windows (MSVC).

For a build with all components enabled in a release configuration the following CMake commands will configure and build the project in a `build/<preset>` directory. The resulting binaries for Lua can be found in `bin/Release/` and `lib/Release/` subdirectories of that folder.

```sh
cmake --preset <linux|macos|windows>
cmake --build --preset <linux|macos|windows> [--config <Debug|Release>]
```

Installation of the compiled artifacts to a specified target directory can be performed with the following command.

```sh
cmake --build --preset <linux|macos|windows> [--config <Debug|Release>] [--prefix <path to install to>] [--strip]
```

To generate a fully packaged release the following can be executed to create a set of `.tar.xz` and `.zip` files in the build directory for the selected preset.

```sh
cmake --build --preset <linux|macos|windows> [--config <Debug|Release>] --target package
```

## License

This project is based upon the original Lua 5.1 source which is available under the MIT license as documented in the `LICENSE` file at the root of the repository. All modifications atop the original source are also covered under the same original license terms.

## Contributors

- [Daniel "Meorawr" Yates](https://github.com/meorawr)


CHANGELOG.md
# Changelog

## [Unreleased]
### Added
- Added `LUA_DISABLE_LOADLIB` build option to disable the runtime dynamic module loader.

## [v3.1]
### Added
- Added the `pcallwithenv(f, env, [args...])` base library function.
- Added compatibility option interface. This allows toggling any potentially incompatible changes made for reference client compatibility.
  - This is exposed via the debug library as `debug.getcompatopt(name)` and `debug.setcompatopt(name, value)`.
  - Supported option names are currently "setfenv", "gctaint", "gcdebug", and "inerrorhandler" which - if set to 1 - will revert the changes documented below.
### Changed
- The `setfenv` function will no longer allow replacing function environments that have a metatable with an `__environment` key to match new reference client behavior.
- `__gc` metamethods are now invoked with a taint barrier to match new reference client behavior.
- The `debugstack`, `debuglocals`, and `getfenv` functions will now return no results if called by `__gc` metamethods.
- The `debuglocals` function can now be called outside of an error handler.
- Fixed linker errors with inlined security functions in unoptimized builds on non-Windows systems.
- Fixed a correctness issue with `secureexecuterange` where errors in the supplied callback were incorrectly forwarded to the global error handler.
- Fixed an issue with `secureexecuterange` where the the C stack would grow each time the callback errored.

## [v3.0]
### Added
- Added a build option (`LUA_USE_FAST_MATH`) to toggle the addition of floating point optimization compiler flags, which have been removed from the presets. This is enabled by default.
- Added a new `n` parameter to `lua_cleartaint` which will clear the taint off the top `n` stack variables.
- Added a backport of `luaL_tolstring` from Lua 5.2.
### Changed
- Fixed a few issues with the `bit.*` library functions.
- Required CMake version is now 3.24.
- CMake will now consult find_package modules to search for dependency libraries before falling back to acquiring them via FetchContent.
### Removed
- Removed `lua_istaintexpected`. This API was effectively non-functional when called by C due to the internals of taint and would always return 1.

## [v2.2]
### Added
- Added `debuglocals([level])` library API.
- Added `debugstack([thread, ][level, ntop, nbase, ...])` library API.
- Added `strsplittable(delim, str[, limit])` library API.
- Added support for object name replacements in handled errors.
### Changed
- Fixed an issue where `secureexecuterange()` would not pass through additional arguments to the supplied function.
- Fixed an issue where `strsplit()` would ignore the value supplied to the `limit` parameter.
- Fixed various correctness issues with `strsplit()` and processing of parameters with embedded null characters.

## [v2.1]
### Changed
- The `lua_Clock` typedef is now an `int64_t` to work around issues on 32-bit platforms.
- Fixed a bug with `table.concat` and `unpack` using an incorrect default value for the optional `j` parameter.

## [v2.0]
### Added
- Reimplemented taint propagation from the ground-up, fixing a large amount of bugs and unimplemented aspects particularly around coroutines.
- Added C and Lua APIs for creating and testing C closures.
- Added C and Lua APIs for profiling and memory measurement.
- Added C and Lua APIs for high-precision timers.
- Added C and Lua APIs for querying and manipulating taint.
- Added C and Lua APIs for secure random number generation.
- Added support for the `__environment` metatable key to `getfenv`.
- Added a `-p` flag to the interpreter to enable performance accounting.
- Added a `-t` flag to the interpreter to run inline scripts (`-e`) or load files insecurely.
  - When this flag is specified, any modules loaded via `-l` will still be loaded securely.

### Changed
- Build process now requires CMake 3.22 and makes use of CMake presets for common build targets. See README for more information.

### Removed
- Removed the taint field on the `lua_Debug` to fix potential binary incompatibilities.
- Removed the `"s"` field from `debug.getinfo`. Use `debug.getstacktaint()` instead.
- Removed `debug.forcesecure()`. Use `debug.setstacktaint(nil)` instead.

## [v1]
### Added
- Added basic taint propagation system.
- Added the following security extensions to the base library:
  - `forceinsecure()`
  - `geterrorhandler()`
  - `hooksecurefunc([table,] "name", func)`
  - `issecure()`
  - `issecurevariable([table,] "variable")`
  - `scrub(...)`
  - `seterrorhandler(errfunc)`

lua/lua.c
/* Licensed under the terms of the MIT License; see full copyright information
 * in the "LICENSE" file or at <http://www.lua.org/license.html> */

#define lua_c

#include "lua.h"

#include "lauxlib.h"
#include "lualib.h"

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static lua_State *globalL = NULL;

static const char *progname = LUA_PROGNAME;

#if defined(LUA_USE_POSIX)

/*
** Use 'sigaction' when available.
*/
static void setsignal (int sig, void (*handler)(int)) {
    struct sigaction sa;
    sa.sa_handler = handler;
    sa.sa_flags = 0;
    sigemptyset(&sa.sa_mask); /* do not mask any signal */
    sigaction(sig, &sa, NULL);
}

#else

#define setsignal signal

#endif

/*
** Hook set by signal function to stop the interpreter.
*/
static void lstop (lua_State *L, lua_Debug *ar) {
    (void) ar; /* unused arg. */
    lua_sethook(L, NULL, 0, 0); /* reset hook */
    luaL_error(L, "interrupted!");
}

/*
** Function to be called at a C signal. Because a C signal cannot
** just change a Lua state (as there is no proper synchronization),
** this function only sets a hook that, when called, will stop the
** interpreter.
*/
static void laction (int i) {
    int flag = LUA_MASKCALL | LUA_MASKRET | LUA_MASKLINE | LUA_MASKCOUNT;
    setsignal(i, SIG_DFL); /* if another SIGINT happens, terminate process */
    lua_sethook(globalL, lstop, flag, 1);
}

static void print_usage (const char *badoption) {
    luaL_writestringerror("%s: ", progname);
    if (badoption[1] == 'e' || badoption[1] == 'l') {
        luaL_writestringerror("'%s' needs argument\n", badoption);
    } else {
        luaL_writestringerror("unrecognized option '%s'\n", badoption);
    }

    luaL_writestringerror(
        "usage: %s [options] [script [args]]\n"
        "Available options are:\n"
        "  -e stat    execute string 'stat'\n"
        "  -l name    require library 'name'\n"
        "  -i         enter interactive mode after executing 'script'\n"
        "  -p         enable profiling and statistics collection\n"
        "  -t         load and execute scripts insecurely\n"
        "  -v         show version information\n"
        "  -E         ignore environment variables\n"
        "  -L lua|wow require base library set(s) on startup\n"
        "  -T         disable taint propagation on startup\n"
        "  --         stop handling options\n"
        "  -          stop handling options and execute stdin\n",
        progname);
}

/*
** Prints an error message, adding the program name in front of it
** (if present)
*/
static void l_message (const char *pname, const char *msg) {
    if (pname) {
        luaL_writestringerror("%s: ", pname);
    }
    luaL_writestringerror("%s\n", msg);
}

/*
** Check whether 'status' is not OK and, if so, prints the error
** message on the top of the stack. It assumes that the error object
** is a string, as it was either generated by Lua or by 'msghandler'.
*/
static int report (lua_State *L, int status) {
    if (status != LUA_OK && !lua_isnil(L, -1)) {
        const char *msg = lua_tostring(L, -1);
        l_message(progname, msg);
        lua_pop(L, 1); /* remove message */
    }
    return status;
}

/*
** Message handler used to decorate all errors
*/
static void msghandler (lua_State *L, int allownilerror) {
    if (allownilerror && lua_isnil(L, 1)) {
        return;
    }

    const char *msg = lua_tostring(L, 1);
    if (msg == NULL) { /* is error object not a string? */
        if (luaL_callmeta(L, 1, "__tostring") && /* does it have a metamethod */
            lua_type(L, -1) == LUA_TSTRING) { /* that produces a string? */
            return; /* that is the message */
        } else {
            msg = lua_pushfstring(L, "(error object is a %s value)", luaL_typename(L, 1));
        }
    }
    luaL_traceback(L, L, msg, 1); /* append a standard traceback */
}

/*
** Per-call error handler used to report all errors
*/
static int callerrhandler (lua_State *L) {
    int allownilerror = 1;
    msghandler(L, allownilerror);
    return 1;
}

/*
** Global Error handler used to report all errors
*/
static int globalerrhandler (lua_State *L) {
    int allownilerror = 0;
    msghandler(L, allownilerror);
    l_message(progname, lua_tostring(L, -1));
    return 1;
}

/*
** Interface to 'lua_pcall', which sets appropriate message function
** and C-signal handler. Used to run all chunks.
*/
static int docall (lua_State *L, int narg, int nres) {
    int status;
    int base = lua_gettop(L) - narg; /* function index */
    lua_pushcfunction(L, callerrhandler); /* push message handler */
    lua_insert(L, base); /* put it under function and args */
    globalL = L; /* to be available to 'laction' */
    setsignal(SIGINT, laction); /* set C-signal handler */
    status = lua_pcall(L, narg, nres, base);
    setsignal(SIGINT, SIG_DFL); /* reset C-signal handler */
    lua_remove(L, base); /* remove message handler from the stack */
    return status;
}

static void print_version (void) {
    static const char lua_version[] = LUA_RELEASE "  " LUA_COPYRIGHT;
    static const char elune_version[] = ELUNE_RELEASE "  " ELUNE_COPYRIGHT;

    luaL_writestring(lua_version, strlen(lua_version));
    luaL_writeline();
    luaL_writestring(elune_version, strlen(elune_version));
    luaL_writeline();
}

/*
** Create the 'arg' table, which stores all arguments from the
** command line ('argv'). It should be aligned so that, at index 0,
** it has 'argv[script]', which is the script name. The arguments
** to the script (everything after 'script') go to positive indices;
** other arguments (before the script name) go to negative indices.
** If there is no script name, assume interpreter's name as base.
*/
static void createargtable (lua_State *L, char **argv, int argc, int script) {
    int i;
    int narg;
    if (script == argc) {
        script = 0; /* no script name? */
    }
    narg = argc - (script + 1); /* number of positive indices */
    lua_createtable(L, narg, script + 1);
    for (i = 0; i < argc; i++) {
        lua_pushstring(L, argv[i]);
        lua_rawseti(L, -2, i - script);
    }
    lua_setglobal(L, "arg");
}

static int dochunk (lua_State *L, int status) {
    if (status == LUA_OK) {
        status = docall(L, 0, 0);
    }
    return report(L, status);
}

static int dofile (lua_State *L, const char *name) {
    return dochunk(L, luaL_loadfile(L, name));
}

static int dostring (lua_State *L, const char *s, const char *name) {
    return dochunk(L, luaL_loadbuffer(L, s, strlen(s), name));
}

/*
** Receives 'globname[=modname]' and runs 'globname = require(modname)'.
*/
static int dolibrary (lua_State *L, char *globname) {
    int status;
    char *modname = strchr(globname, '=');
    if (modname == NULL) { /* no explicit name? */
        modname = globname; /* module name is equal to global name */
    } else {
        *modname = '\0'; /* global name ends here */
        modname++; /* module name starts after the '=' */
    }
    lua_getglobal(L, "require");
    lua_pushstring(L, modname);
    status = docall(L, 1, 1); /* call 'require(modname)' */
    if (status == LUA_OK) {
        lua_setglobal(L, globname); /* globname = require(modname) */
    }
    return report(L, status);
}

/*
** Push on the stack the contents of table 'arg' from 1 to #arg
*/
static int pushargs (lua_State *L) {
    int i;
    int n;
    lua_getglobal(L, "arg");
    if (lua_type(L, -1) != LUA_TTABLE) {
        luaL_error(L, "'arg' is not a table");
    }
    n = (int) lua_objlen(L, -1);
    luaL_checkstack(L, n + 3, "too many arguments to script");
    for (i = 1; i <= n; i++) {
        lua_rawgeti(L, -i, i);
    }
    lua_remove(L, -i); /* remove table from the stack */
    return n;
}

static int handle_script (lua_State *L, char **argv) {
    int status;
    const char *fname = argv[0];
    if (strcmp(fname, "-") == 0 && strcmp(argv[-1], "--") != 0) {
        fname = NULL; /* stdin */
    }
    status = luaL_loadfile(L, fname);
    if (status == LUA_OK) {
        int n = pushargs(L); /* push arguments to script */
        status = docall(L, n, LUA_MULTRET);
    }
    return report(L, status);
}

/* bits of various argument indicators in 'args' */
enum {
    has_error = 1, /* bad option */
    has_i = 2, /* -i */
    has_v = 4, /* -v */
    has_e = 8, /* -e */
    has_E = 16, /* -E */
    has_p = 32, /* -p */
    has_T = 64, /* -T */
    has_L = 128, /* -L */
};

/*
** Traverses all arguments from 'argv', returning a mask with those
** needed before running any Lua code (or an error code if it finds
** any invalid argument). 'first' returns the first not-handled argument
** (either the script name or a bad argument in case of error).
*/
static int collectargs (char **argv, int *first) {
    int args = 0;
    int i;
    for (i = 1; argv[i] != NULL; i++) {
        *first = i;
        if (argv[i][0] != '-') { /* not an option? */
            return args; /* stop handling options */
        }
        switch (argv[i][1]) { /* else check option */
            case '-': /* '--' */
                if (argv[i][2] != '\0') { /* extra characters after '--'? */
                    return has_error; /* invalid option */
                }
                *first = i + 1;
                return args;
            case '\0': /* '-' */
                return args; /* script "name" is '-' */
            case 'E':
                if (argv[i][2] != '\0') { /* extra characters? */
                    return has_error; /* invalid option */
                }
                args |= has_E;
                break;
            case 'p':
                if (argv[i][2] != '\0') { /* extra characters? */
                    return has_error; /* invalid option */
                }
                args |= has_p;
                break;
            case 't':
                if (argv[i][2] != '\0') { /* extra characters? */
                    return has_error; /* invalid option */
                }
                break;
            case 'T':
                if (argv[i][2] != '\0') { /* extra characters? */
                    return has_error; /* invalid option */
                }
                args |= has_T;
                break;
            case 'i':
                args |= has_i; /* (-i implies -v) */
                LUA_FALLTHROUGH;
            case 'v':
                if (argv[i][2] != '\0') { /* extra characters? */
                    return has_error; /* invalid option */
                }
                args |= has_v;
                break;
            case 'e':
                args |= has_e;
                goto check_has_argument;
            case 'L':
                args |= has_L;
                goto check_has_argument;
            case 'l': /* all three options need an argument */
            check_has_argument:
                if (argv[i][2] == '\0') { /* no concatenated argument? */
                    i++; /* try next 'argv' */
                    if (argv[i] == NULL || argv[i][0] == '-') {
                        return has_error; /* no next argument or it is another option */
                    }
                }
                break;
            default: /* invalid option */
                return has_error;
        }
    }
    *first = i; /* no script name */
    return args;
}

/*
** Processes option 'L' to load base libraries for the Lua state.
*/
static int libargs (lua_State *L, char **argv, int n) {
    int i;
    int loaded = 0;

    for (i = 1; i < n; i++) {
        int option = argv[i][1];
        char *extra = NULL;

        lua_assert(argv[i][0] == '-'); /* already checked */

        if (option == 'L' || option == 'e' || option == 'l') {
            extra = argv[i] + 2; /* all options need an argument */

            if (*extra == '\0') {
                extra = argv[++i];
            }
        }

        if (option == 'L') {
            lua_assert(extra != NULL);

            if (strcmp(extra, "lua") == 0) {
                luaL_openlibsx(L, LUALIB_STANDARD);
                loaded = 1;
            } else if ((strcmp(extra, "elune") == 0) || strcmp(extra, "wow") == 0) {
                luaL_openlibsx(L, LUALIB_ELUNE);
                loaded = 1;
            } else if (strcmp(extra, "none") == 0) {
                loaded = 1;
            } else {
                lua_pushfstring(L, "unknown base library type: %s", extra);
                report(L, LUA_ERRRUN);
                return 0;
            }
        }
    }

    if (!loaded) {
        /* Load both libraries with standard taking priority by default. */
        luaL_openlibsx(L, LUALIB_ELUNE);
        luaL_openlibsx(L, LUALIB_STANDARD);
    }

    return 1;
}

/*
** Processes options 'e', 'l', and 't', which involve running Lua code.
** Returns 0 if some code raises an error.
*/
static int runargs (lua_State *L, char **argv, int n) {
    int i;
    for (i = 1; i < n; i++) {
        int option = argv[i][1];
        lua_assert(argv[i][0] == '-'); /* already checked */
        switch (option) {
            case 'L':
            case 'e':
            case 'l': {
                int status;
                char *extra = argv[i] + 2; /* all options need an argument */
                if (*extra == '\0') {
                    extra = argv[++i];
                }
                lua_assert(extra != NULL);
                if (option == 'e') {
                    status = dostring(L, extra, "=(command line)");
                } else if (option == 'l') {
                    status = dolibrary(L, extra);
                } else {
                    status = LUA_OK; /* ignored argument ('-L') */
                }
                if (status != LUA_OK) {
                    return 0;
                }
                break;
            }
            case 't': {
                luaL_forceinsecure(L);
                break;
            }
        }
    }
    return 1;
}

static int handle_luainit (lua_State *L) {
    const char *name = "=" LUA_INIT;
    const char *init = getenv(name + 1);
    if (init == NULL) {
        return LUA_OK;
    } else if (init[0] == '@') {
        return dofile(L, init + 1);
    } else {
        return dostring(L, init, name);
    }
}

/*
** {==================================================================
** Read-Eval-Print Loop (REPL)
** ===================================================================
*/

/*
** Return the string to be used as a prompt by the interpreter. Leave
** the string (or nil, if using the default value) on the stack, to keep
** it anchored.
*/
static const char *get_prompt (lua_State *L, int firstline) {
    lua_getglobal(L, firstline ? "_PROMPT" : "_PROMPT2");
    if (lua_type(L, -1) == LUA_TNIL) {
        return (firstline ? LUA_PROMPT : LUA_PROMPT2); /* use the default */
    } else { /* apply 'tostring' over the value */
        const char *p = lua_tolstring(L, -1, NULL);
        lua_remove(L, -2); /* remove original value */
        return p;
    }
}

/* mark in error messages for incomplete statements */
#define EOFMARK "'<eof>'"
#define marklen (sizeof(EOFMARK) / sizeof(char) - 1)

/*
** Check whether 'status' signals a syntax error and the error
** message at the top of the stack ends with the above mark for
** incomplete statements.
*/
static int incomplete (lua_State *L, int status) {
    if (status == LUA_ERRSYNTAX) {
        size_t lmsg;
        const char *msg = lua_tolstring(L, -1, &lmsg);
        if (lmsg >= marklen && strcmp(msg + lmsg - marklen, EOFMARK) == 0) {
            lua_pop(L, 1);
            return 1;
        }
    }
    return 0; /* else... */
}

/*
** Prompt the user, read a line, and push it into the Lua stack.
*/
static int pushline (lua_State *L, int firstline) {
    const char *prompt = get_prompt(L, firstline);
    const char *line;

    if (luaL_readline(L, prompt) != 0) {
        return 0; /* no input (prompt will be popped by caller) */
    }

    lua_remove(L, -2); /* remove prompt */

    line = lua_tostring(L, -1);
    if (firstline && line[0] == '=') { /* for compatibility with 5.2, ... */
        lua_pushfstring(L, "return %s", line + 1); /* change '=' to 'return' */
        lua_remove(L, -2); /* remove original input */
    }

    return 1;
}

/*
** Try to compile line on the stack as 'return <line>;'; on return, stack
** has either compiled chunk or original line (if compilation failed).
*/
static int addreturn (lua_State *L) {
    const char *line = lua_tostring(L, -1); /* original line */
    const char *retline = lua_pushfstring(L, "return %s;", line);
    int status = luaL_loadbuffer(L, retline, strlen(retline), "=stdin");
    if (status == LUA_OK) {
        lua_remove(L, -2); /* remove modified line */
        if (line[0] != '\0') { /* non empty? */
            luaL_saveline(L, line); /* keep history */
        }
    } else {
        lua_pop(L, 2); /* pop result from 'luaL_loadbuffer' and modified line */
    }
    return status;
}

/*
** Read multiple lines until a complete Lua statement
*/
static int multiline (lua_State *L) {
    for (;;) { /* repeat until gets a complete statement */
        size_t len;
        const char *line = lua_tolstring(L, 1, &len); /* get what it has */
        int status = luaL_loadbuffer(L, line, len, "=stdin"); /* try it */
        if (!incomplete(L, status) || !pushline(L, 0)) {
            luaL_saveline(L, line); /* keep history */
            return status; /* cannot or should not try to add cont. line */
        }
        lua_pushliteral(L, "\n"); /* add newline... */
        lua_insert(L, -2); /* ...between the two lines */
        lua_concat(L, 3); /* join them */
    }
}

/*
** Read a line and try to load (compile) it first as an expression (by
** adding "return " in front of it) and second as a statement. Return
** the final status of load/call with the resulting function (if any)
** in the top of the stack.
*/
static int loadline (lua_State *L) {
    int status;
    lua_settop(L, 0);
    if (!pushline(L, 1)) {
        return -1; /* no input */
    }
    if ((status = addreturn(L)) != LUA_OK) { /* 'return ...' did not work? */
        status = multiline(L); /* try as command, maybe with continuation lines */
    }
    lua_remove(L, 1); /* remove line from the stack */
    lua_assert(lua_gettop(L) == 1);
    return status;
}

/*
** Prints (calling the Lua 'print' function) any values on the stack
*/
static void l_print (lua_State *L) {
    int n = lua_gettop(L);
    if (n > 0) { /* any result to be printed? */
        luaL_checkstack(L, LUA_MINSTACK, "too many results to print");
        lua_getglobal(L, "print");
        lua_insert(L, 1);
        if (lua_pcall(L, n, 0, 0) != LUA_OK) {
            l_message(progname, lua_pushfstring(L, "error calling 'print' (%s)", lua_tostring(L, -1)));
        }
    }
}

/*
** Do the REPL: repeatedly read (load) a line, evaluate (call) it, and
** print any results.
*/
static void doREPL (lua_State *L) {
    int status;
    const char *oldprogname = progname;
    progname = NULL; /* no 'progname' on errors in interactive mode */
    luaL_setreadlinename(L, "lua");
    while ((status = loadline(L)) != -1) {
        if (status == LUA_OK) {
            status = docall(L, 0, LUA_MULTRET);
        }
        if (status == LUA_OK) {
            l_print(L);
        } else {
            report(L, status);
        }
    }
    lua_settop(L, 0); /* clear stack */
    luaL_writeline();
    progname = oldprogname;
}

/* }================================================================== */

/*
** Main body of stand-alone interpreter (to be called in protected mode).
** Reads the options and handles them all.
*/
static int pmain (lua_State *L) {
    int argc = (int) lua_tointeger(L, 1);
    char **argv = (char **) lua_touserdata(L, 2);
    int script;
    int args = collectargs(argv, &script);
    if (argv[0] && argv[0][0]) {
        progname = argv[0];
    }
    if (args == has_error) { /* bad arg? */
        print_usage(argv[script]); /* 'script' has index of bad arg. */
        return 0;
    }
    if (args & has_v) { /* option '-v'? */
        print_version();
    }
    if (!(args & has_T)) { /* option '-T' not specified? */
        lua_settaintmode(L, LUA_TAINTRDRW);
    }
    if (args & has_p) { /* option '-p'? */
        lua_setprofilingenabled(L, 1);
    }
    if (args & has_E) { /* option '-E'? */
        lua_pushboolean(L, 1); /* signal for libraries to ignore env. vars. */
        lua_setfield(L, LUA_REGISTRYINDEX, "LUA_NOENV");
    }
    if (!libargs(L, argv, script)) { /* execute argument -L */
        return 0; /* something failed */
    }
    createargtable(L, argv, argc, script); /* create table 'arg' */
    lua_pushcclosure(L, globalerrhandler, 0);
    lua_replace(L, LUA_ERRORHANDLERINDEX); /* install error handler */
    if (!(args & has_E)) { /* no option '-E'? */
        if (handle_luainit(L) != LUA_OK) { /* run LUA_INIT */
            return 0; /* error running LUA_INIT */
        }
    }
    if (!runargs(L, argv, script)) { /* execute arguments -e and -l */
        return 0; /* something failed */
    }
    if (script < argc && /* execute main script (if there is one) */
        handle_script(L, argv + script) != LUA_OK) {
        return 0;
    }
    if (args & has_i) { /* -i option? */
        doREPL(L); /* do read-eval-print loop */
    } else if (script == argc && !(args & (has_e | has_v))) { /* no arguments? */
        if (luaL_stdinistty(L)) { /* running in interactive mode? */
            print_version();
            doREPL(L); /* do read-eval-print loop */
        } else {
            dofile(L, NULL); /* executes stdin as a file */
        }
    }
    lua_pushboolean(L, 1); /* signal no errors */
    return 1;
}

int main (int argc, char **argv) {
    int status;
    int result;
    lua_State *L = luaL_newstate(); /* create state */
    if (L == NULL) {
        l_message(argv[0], "cannot create state: not enough memory");
        return EXIT_FAILURE;
    }
    lua_pushcfunction(L, &pmain); /* to call 'pmain' in protected mode */
    lua_pushinteger(L, argc); /* 1st argument */
    lua_pushlightuserdata(L, argv); /* 2nd argument */
    status = lua_pcall(L, 2, 1, 0); /* do the call */
    result = lua_toboolean(L, -1); /* get result */
    report(L, status);
    lua_close(L);
    return (result && status == LUA_OK) ? EXIT_SUCCESS : EXIT_FAILURE;
}
